src/daft-file/src/file.rs

```diff
@@ -28,38 +28,46 @@ impl TryFrom<FileReference> for DaftFile {
         let rt = common_runtime::get_io_runtime(true);
 
-        let (source, path, file_size) = rt.block_within_async_context(async move {
-            let (source, path) = io_client
-                .get_source_and_path(&path)
-                .await
-                .map_err(DaftError::from)?;
-            // getting the size is pretty cheap, so we do it upfront
-            // we grab the size upfront so we can use it to determine if we are at the end of the file
-            let file_size = source
-                .get_size(&path, None)
-                .await
-                .map_err(|e| DaftError::ComputeError(e.to_string()))?;
-            DaftResult::Ok((source, path, file_size))
-        })??;
-
-        // Default to 8MB buffer
-        const DEFAULT_BUFFER_SIZE: usize = 8 * 1024 * 1024;
-
-        let reader = ObjectSourceReader::new(source, path, None, file_size, io_config);
-
-        // we wrap it in a BufReader so we are not making so many network requests for each byte read
-        let buffered_reader = BufReader::with_capacity(DEFAULT_BUFFER_SIZE, reader);
+        let (source, path, file_size, supports_range) =
+            rt.block_within_async_context(async move {
+                let (source, path) = io_client
+                    .get_source_and_path(&path)
+                    .await
+                    .map_err(DaftError::from)?;
+                // getting the size is pretty cheap, so we do it upfront
+                // we grab the size upfront so we can use it to determine if we are at the end of the file
+                let file_size = source
+                    .get_size(&path, None)
+                    .await
+                    .map_err(|e| DaftError::ComputeError(e.to_string()))?;
+                let supports_range = source
+                    .supports_range(&path)
+                    .await
+                    .map_err(|e| DaftError::ComputeError(e.to_string()))?;
+                DaftResult::Ok((source, path, file_size, supports_range))
+            })??;
+
+        // Default to 16MB buffer
+        const DEFAULT_BUFFER_SIZE: usize = 16 * 1024 * 1024;
+
+        let mut reader = ObjectSourceReader::new(source, path, None, file_size);
+        if !supports_range || file_size <= DEFAULT_BUFFER_SIZE {
+            let mut buf = Vec::with_capacity(file_size);
+            reader.read_to_end(&mut buf)?;
+            Ok(Self::from_bytes(buf))
+        } else {
+            // we wrap it in a BufReader so we are not making so many network requests for each byte read
+            let buffered_reader = BufReader::with_capacity(DEFAULT_BUFFER_SIZE, reader);
 
-        Ok(Self {
-            cursor: Some(FileCursor::ObjectReader(buffered_reader)),
-            position: 0,
-        })
+            Ok(Self {
+                cursor: Some(FileCursor::ObjectReader(buffered_reader)),
+                position: 0,
+            })
+        }
```

Comment:
These changes don't seem to be related to making daft.File immutable. What are they for?